Process inner html of blocks when escaping text content

jffng opened the pull request:
**What**
This PR enables HTML markup inside block content to be processed while still allowing the text content to be safely escaped.

Before | After
--- | ---
<video src="https://github.com/user-attachments/assets/a9c1a925-5622-43cc-ace3-5a2942f5b936"> | <video src="https://github.com/user-attachments/assets/367f866e-3f9e-4cc5-b781-0b336e19dc5f">

**How**
Updates the `CBT_Theme_Locale::escape_text_content` function to handle nested HTML markup. 

Uses `WP_HTML_Tag_Processor` to process every token of the blocks content, and generates a string that can be formatted and then translated. 

**Why**
Fixes #573 #682 and #691.  

**To test**
- Add some markup to the inside of a block in a template.
- Save the changes to the theme.
- Verify the template has been patternized correctly, and the template appears as expected.

jffng committed:
WIP working to convert attributes and URLs.

jffng committed:
Ensure media is added to local.

jffng committed:
Use a list of tokens and sprintf to generate the formatted string.

jffng committed:
Only format the string if tokens are present.

jffng committed:
Update content with html test.

Update remaining tests with inner markup.

Try and fix tests.

Format tests.

jffng committed:
Remove whitespace.

jffng commented on includes/create-theme/theme-locale.php:
```diff
@@ -28,6 +28,72 @@ private static function escape_text_content( $string ) {
+					$tokens[] = "</{$token_name}>";
+				} else {
+					// Depending on the HTML tag, we may need to process attributes so they are correctly added to the placeholder.
+					switch ( $token_name ) {
```
One thing I would like feedback on is whether this is too fragile, complicated, or tedious an approach to maintain.

I plan to add a few more unit tests.

mikachan replied:
I think we're good to rely on the output of `get_token_name()`, so based on that, I believe this is a fine way to parse these tags. We have a limited scope since we're currently only dealing with `a`, `img` and `mark`, and otherwise we default to the original tag.

That's only my opinion though 😅 Was there anything you were specifically concerned about with this approach?

jffng replied:
I think just needing to add cases for specific tags and if the attributes changes, it may break. 

But the options for adding inline HTML / formatting inside blocks are rather limited, so maybe it's okay.

creativecoder replied:
Do we have to deconstruct the attributes and reassemble them? I'm worried about dropping attributes or other parts of the markup that might be added by plugins or filters.

For example, could we replace `<a href="https://wordpress.org">WordPress</a>` with something like
```php
<a href="<?php echo esc_url( 'https://wordpress.org' ) >?"><?php echo esc_html__( 'WordPress' ) ?></a>
``` 
but maintain any other attributes added to the `<a>` tag?

I'm not really worried about escaping every possible attribute, I think the main thing is to make sure the translated strings are escaped as a form of user input.

jffng replied:
> I'm worried about dropping attributes or other parts of the markup that might be added by plugins or filters.

That's a good point. I could not find a way with the HTML Tag Processor to just carry over the whole tag token and all its attributes. But I refactored the approach so that all the attributes should be carried over, rather than processing the attributes based on the type of tag: https://github.com/WordPress/create-block-theme/pull/719/commits/678a8ef22ddecc2e0b4190dd141cfa0c4436cd1e#diff-fc11ad66397ed68725a8fe74f3031f104368f850b5b94370123482ec9e719e1bR54-R70

madhusudhand reviewed (commented):
It works! But I encountered following scenario where multiple individually translated strings now becomes the single translation.

I guess this is because the template in the editor comes with a single string (after translations) and now it is seen as a single string. 

I don't think this should be blocking this change. It is to be addressed separately by reading original pattern and current template together may be?

**Original**
```
<h2 class="wp-block-heading" id="botswana-new-zealand-south-korea-japan-madagascar" style="font-size:80px;font-style:normal;font-weight:900;line-height:1.1;text-transform:uppercase"><a href=""><?php echo esc_html__( 'ITALY', 'adventurer' ); ?></a><br><a href=""><?php echo esc_html__( 'COSTA RICA', 'adventurer' ); ?></a><br><a href=""><?php echo esc_html__( 'CANADA', 'adventurer' ); ?></a><br><a href=""><?php echo esc_html__( 'LAOS', 'adventurer' ); ?></a><br><a href=""><?php echo esc_html__( 'TURKEY', 'adventurer' ); ?></a></h2>
```

**After CBT localization**
```
<h2 class="wp-block-heading" id="botswana-new-zealand-south-korea-japan-madagascar" style="font-size:80px;font-style:normal;font-weight:900;line-height:1.1;text-transform:uppercase"><?php /* Translators: %s are html tags */ echo sprintf( esc_html__( '%sITALY%s%s%sCOSTA RICA%s%s%sCANADA%s%s%sLAOS%s%s%sTURKEY%s', 'adventurer' ), '<a href="">', '</a>', '<br>', '<a href="">', '</a>', '<br>', '<a href="">', '</a>', '<br>', '<a href="">', '</a>', '<br>', '<a href="">', '</a>' ); ?></h2>
```

mikachan reviewed (commented):
Thanks so much for handling this @jffng! I can see this will fix the issues we've been seeing with translations ✨ 

@creativecoder, if you get a chance it would be great to hear your thoughts on this.

creativecoder commented on includes/create-theme/theme-locale.php:
```diff
@@ -28,6 +28,72 @@ private static function escape_text_content( $string ) {
 
 		$string = addcslashes( $string, "'" );
 
```
Perhaps there's enough going on here this should have it's own utility class?

jffng replied:
I tried splitting this out in [bf8ff55](https://github.com/WordPress/create-block-theme/pull/719/commits/bf8ff55e9629f9bf31317bf18e806f5d086b61ef).

creativecoder commented on includes/create-theme/theme-locale.php:
```diff
@@ -28,6 +28,72 @@ private static function escape_text_content( $string ) {
+							$style    = esc_attr( $p->get_attribute( 'style' ) );
+							$class    = esc_attr( $p->get_attribute( 'class' ) );
+							$tokens[] = "<mark style=\"{$style}\" class=\"{$class}\">";
+							break;
```
I think we should be printing out `esc_` functions here, rather than executing them. Typically the advice is to escape as late as possible, which means we should see those functions in the PHP output.

jffng replied:
Done here: https://github.com/WordPress/create-block-theme/pull/719/commits/678a8ef22ddecc2e0b4190dd141cfa0c4436cd1e

creativecoder commented on includes/create-theme/theme-locale.php:
```diff
@@ -28,6 +28,72 @@ private static function escape_text_content( $string ) {
+		// If tokens is not empty, format the string using sprintf.
+		if ( ! empty( $tokens ) ) {
+			// Format the string, replacing the placeholders with the formatted tokens.
+			return "<?php /* Translators: %s are html tags */ echo sprintf( esc_html__( '$text', '" . wp_get_theme()->get( 'TextDomain' ) . "' ), " . implode(
```
`Translators: %s are html tags` doesn't seem like a very helpful comment. From what I understand, the best practices for using multiple placeholders includes

- Use numbered placeholders (`%1$s` rather than `%s`)
- Name the tag in the comment: (e.g. `/* translators: %1$s: start of link (<a>), %2$s: end of link (</a>) */`)

madhusudhand replied:
If the content itself is having a `%s`, it breaks the page because of the improper closings.
Replacing such instances with `%%s` or `%%1$s` will fix things.

jffng replied:
> If the content itself is having a %s, it breaks the page because of the improper closings.

Good catch, I fixed it here: 3f0adfac69e3a240e58257b1bbe40d1db8bef9a4

creativecoder commented on includes/create-theme/theme-locale.php:
```diff
@@ -28,6 +28,72 @@ private static function escape_text_content( $string ) {
+					$tokens
+				)
+			) . ' ); ?>';
+		}
```
If there's a line break (`<br>`), that seems like an indicator to break the content there to create a separate string. I think that will help avoid the situation @madhusudhand [found](https://github.com/WordPress/create-block-theme/pull/719#pullrequestreview-2300097301).

jffng replied:
I'm not sure we can or should assume this, see for example the screen recording and use of `<br>` in issue https://github.com/WordPress/create-block-theme/issues/682.

creativecoder reviewed (commented):
Thanks for trying this; it looks like a very good start!

madhusudhand commented on tests/test-theme-templates.php:
```diff
@@ -262,7 +262,12 @@ public function test_localize_verse() {
+		<!-- /wp:verse -->';
+
+		$this->assertStringContainsString( $expected_output, $new_template->content );
 	}
```
Let's add a test case for this `If the content itself is having a %s, it breaks the page because of the improper closings.`

jffng replied:
Done here: [4504111](https://github.com/WordPress/create-block-theme/pull/719/commits/4504111b0900b9d377984ab2794a297f94a9659e)

jffng committed:
Process all attributes
Escape at the end.
Provide better translation note.

jffng committed:
Refactor token processing to its own class.

jffng committed:
Update tests with string replacements and translation.

jffng committed:
Check if % exists in the text and escape it.

jffng committed:
Add a test case for a localizing text that includes a %.

jffng commented:
I think I addressed all the feedback:

- Changed approach so we don't drop any attributes
- Refactored to its own class.
- Handled case when text contains `%` and added a test for it

Ready for another review.

mikachan reviewed (approved):
Thanks so much @jffng! I've tested the latest changes out and I believe everything is working as described. Great job 👏 

Here are my test results, using the Adventurer theme as my test theme, and editing the Blog Home template (which uses a list of headings with links):

| trunk | This PR |
| ----- | ------ |
| ![image](https://github.com/user-attachments/assets/a9813f05-1bb5-4fb0-9821-351812c2ee8d) | ![image](https://github.com/user-attachments/assets/366b2195-a578-432e-ab53-08e2037dd508) |

creativecoder commented on includes/create-theme/theme-locale.php:
```diff
@@ -28,6 +31,24 @@ private static function escape_text_content( $string ) {
+		$translators_note = $p->get_translators_note();
+
+		if ( ! empty( $tokens ) ) {
+			return "<?php $translators_note echo sprintf( esc_html__( '$text', '" . wp_get_theme()->get( 'TextDomain' ) . "' ), " . implode(
```
I'm not certain, but I suspect the translators comment may need to be on a separate line, immediately above the `esc_html__` function, to be parsed correctly.

mikachan replied:
This should be addressed in https://github.com/WordPress/create-block-theme/pull/719/commits/1e317ce80b0ef55f28dccad825264b5410434f9f.

creativecoder commented on includes/create-theme/theme-token-processor.php:
```diff
@@ -0,0 +1,127 @@
+
+					$token                  .= '>';
+					$this->tokens[]          = $token;
+					$this->translators_note .= '%' . $this->increment . "\$s is the start of a '" . $token_name . "' HTML element";
```
Do we have any way of differentiating self closing tags? For example "%3$s is the start of a 'br' HTML element" might be a little confusing, since it doesn't (usually) have a closing tag.

mikachan replied:
Handled in https://github.com/WordPress/create-block-theme/pull/719/commits/6b5fe21f2372f6918cb0e1f6e825b7721d44cfab.

creativecoder reviewed (commented):
This is working well at solving the problem in my testing! 🎉

Here's the block markup I used:

```html
<!-- wp:paragraph {"align":"center"} -->
<p class="has-text-align-center"><strong>Études</strong> is a pioneering firm that seamlessly merges creativity<br>and functionality to <a href="https://wordpress.org" data-type="link" data-id="https://wordpress.org">redefine <em>architectural</em> excellence</a>.</p>
<!-- /wp:paragraph -->
```

And here's the output in the .php pattern file:

```php
<?php /* Translators: %1$s is the start of a 'strong' HTML element, %2$s is the end of a 'strong' HTML element, %3$s is the start of a 'br' HTML element, %4$s is the start of a 'a' HTML element, %5$s is the start of a 'em' HTML element, %6$s is the end of a 'em' HTML element, %7$s is the end of a 'a' HTML element */  echo sprintf( esc_html__( '%1$sÉtudes%2$s is a pioneering firm that seamlessly merges creativity%3$sand functionality to %4$sredefine %5$sarchitectural%6$s excellence%7$s.', 'twentytwentyfour' ), '<strong>', '</strong>', '<br>', '<a href="' . esc_url( 'https://wordpress.org' ) . '" data-type="link" data-id="https://wordpress.org">', '<em>', '</em>', '</a>' ); ?>
```

A few changes I think we need to make this output correct

- I believe the `/* translators:` comment needs to be on a separate line, immediately above the call to the translation function (but I'm not 100% sure)
- When there are multiple placeholders in a string, the comment should be numbered, e.g. `/* Translators: 1: start of 'strong' HTML element, 2: ...` ([see plugin docs](https://developer.wordpress.org/plugins/internationalization/how-to-internationalize-your-plugin/#variables))
- If there's a way to differentiate between tags that are self closing (like `<br>`), it would be nice to modify the comment (`1: 'br' HTML element, 2: ...`), so it's not confusing when there's no corresponding closing tag

mikachan committed:
Add new line at the end of translators note

mikachan committed:
Reformat numbers in translators note

mikachan committed:
Update test

mikachan committed:
Move new line to theme-locale

mikachan committed:
Handle self closing tags

mikachan committed:
Refactor how new line is added

mikachan committed:
Attempt to fix tests

mikachan committed:
Attempt to fix tests again

mikachan commented:
Thanks for all the feedback on this one!

I think I've addressed this feedback:

> I believe the /* translators: comment needs to be on a separate line, immediately above the call to the translation function (but I'm not 100% sure)

Done in https://github.com/WordPress/create-block-theme/pull/719/commits/1e317ce80b0ef55f28dccad825264b5410434f9f.

> When there are multiple placeholders in a string, the comment should be numbered, e.g. /* Translators: 1: start of 'strong' HTML element, 2: ... ([see plugin docs](https://developer.wordpress.org/plugins/internationalization/how-to-internationalize-your-plugin/#variables))

Done in https://github.com/WordPress/create-block-theme/pull/719/commits/c6652860bd350295ad7c85a00033936b132308a1.

> If there's a way to differentiate between tags that are self closing (like <br>), it would be nice to modify the comment (1: 'br' HTML element, 2: ...), so it's not confusing when there's no corresponding closing tag

Done in https://github.com/WordPress/create-block-theme/commit/6b5fe21f2372f6918cb0e1f6e825b7721d44cfab.

Would appreciate any reviews on this so we can hopefully land this soon 🙇

matiasbenedetto committed:
Merge branch 'trunk' into fix/nested-html-translation-string-conversion

matiasbenedetto reviewed (approved):
This is testing fine for me :+1: 
I think it's good enough to solve the translation needs of theme builders.
Let's merge.

annezazu commented:
Thank you all so much for getting this done!

iamtakashi commented:
Nice! Thank you so much for the fix.
